File: src/main/java/dev/bypixel/redivelocity/listeners/ServerSwitchListener.java
/*
 * Copyright (c) 2025.
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package dev.bypixel.redivelocity.listeners;

import com.velocitypowered.api.event.Subscribe;
import com.velocitypowered.api.event.player.ServerConnectedEvent;
import dev.bypixel.redivelocity.RediVelocity;
import dev.bypixel.redivelocity.RediVelocityLogger;
import dev.bypixel.redivelocity.config.Config;
import dev.bypixel.redivelocity.jedisWrapper.RedisController;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;
import net.kyori.adventure.text.minimessage.MiniMessage;

import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

@Singleton
public class ServerSwitchListener {

    private final RediVelocity rediVelocity;
    private final Config config;
    private final RedisController redisController;
    private final ExecutorService redisExecutor = Executors.newFixedThreadPool(5);
    private final RediVelocityLogger logger;

    @Inject
    public ServerSwitchListener(RediVelocity rediVelocity, Config config, RedisController redisController, RediVelocityLogger logger) {
        this.rediVelocity = rediVelocity;
        this.config = config;
        this.redisController = redisController;
        this.logger = logger;
    }

    @SuppressWarnings("unused")
    @Subscribe
    public void onServerSwitch(ServerConnectedEvent event) {
        var player = event.getPlayer();

        if (Objects.equals(redisController.getString("rv-init-process"), "true")) {
            player.disconnect(MiniMessage.miniMessage().deserialize("<red>Proxy is booting up, please wait..."));
            return;
        }

        var previousServerName = event.getPreviousServer().map(server -> server.getServerInfo().getName()).orElse("null");
        var redisConfig = config.getRedis();

        CompletableFuture.runAsync(() -> {
            redisController.sendServerSwitchMessage(
                    "serverSwitch",
                    rediVelocity.getProxyId(),
                    player.getUsername(),
                    player.getUniqueId().toString(),
                    player.getClientBrand(),
                    player.getRemoteAddress().toString().split(":")[0].substring(1),
                    event.getServer().getServerInfo().getName() != null ? event.getServer().getServerInfo().getName() : "null",
                    previousServerName,
                    redisConfig.getChannel()
            );

            redisController.setHashField("rv-players-server", player.getUniqueId().toString(), event.getServer().getServerInfo().getName());
        }, redisExecutor).exceptionally(ex -> {
            logger.sendErrorLogs("Error while sending server switch Redis message " + ex.getMessage());
            return null;
        });
    }
}

File: src/main/kotlin/dev/bypixel/redivelocity/event/ServerSwitchListener.kt
/*
 * Copyright (c) 2024-present byPixelTV & contributors.
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package dev.bypixel.redivelocity.event

import com.velocitypowered.api.event.Subscribe
import com.velocitypowered.api.event.player.ServerConnectedEvent
import dev.bypixel.redivelocity.RediVelocity
import dev.bypixel.redivelocity.RediVelocityCoroutineScope
import dev.bypixel.redivelocity.feature.globalPlayercount.PlayercountUtil
import io.lettuce.core.ExperimentalLettuceCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.json.JSONObject

object ServerSwitchListener {
    @OptIn(ExperimentalLettuceCoroutinesApi::class)
    @Subscribe
    fun onServerSwitch(event: ServerConnectedEvent) {
        val player = event.player

        val previousServer =
            event.previousServer.map { server -> server.serverInfo.name }.orElse("null")
        val newServer = event.server

        RediVelocityCoroutineScope.launch(Dispatchers.IO) {
            PlayercountUtil.setProxyPlayercount()
            RediVelocity.instance.lettuceClient.sendMessage(JSONObject().apply {
                put("action", "UPDATE")
            }, "redivelocity:global-player-updates")
            RediVelocity.instance.lettuceClient.sendMessage(JSONObject().apply {
                put("action", "SERVER_SWITCH")
                put("uuid", player.uniqueId.toString())
                put("username", player.username)
                put("ip", player.remoteAddress.toString().split(":")[0].substring(1))
                put("proxyId", RediVelocity.instance.proxyId)
                put("protocolVersion", player.protocolVersion.protocol)
                put("clientBrand", player.clientBrand)
                put("timestamp", System.currentTimeMillis())
                put("fromServer", previousServer)
                put("toServer", newServer.serverInfo.name)
            }, "redivelocity:players")

            RediVelocity.instance.lettuceClient.commands.hset("redivelocity:player:servers", player.uniqueId.toString(), newServer.serverInfo.name)
        }
    }
}